project/src/main/java/com/google/sps/data/ProductSearchLibrary.java

```diff
@@ -250,4 +199,94 @@ public static void removeProductFromProductSet(
+            ProductItem productItem = new ProductItem(product.getName().substring(product.getName().lastIndexOf('/') + 1), product.getDisplayName(),
+                                        product.getProductCategory());
+            productItems.add(productItem);
+            for (Product.KeyValue element : product.getProductLabelsList()) {
+                System.out.println(String.format("%s: %s", element.getKey(), element.getValue()));
+            }
+        }
+    } catch(Exception e){
+        System.err.println("Could not list products");
+    }
+    return productItems;
+  }
+   
+   public static ArrayList<String> listReferenceImagesOfProduct(
+    String productId) throws IOException {
+    ArrayList<String> referenceImages = new ArrayList<>();
+    try (ProductSearchClient client = ProductSearchClient.create()) {
+        // Get the full path of the product.
+        String formattedParent =
+            ProductSearchClient.formatProductName(projectId, computeRegion, productId);
+        for (ReferenceImage image : client.listReferenceImages(formattedParent).iterateAll()) {
+            referenceImages.add(image.getUri());
+        }
+    }
+    return referenceImages;
+  }
+
+  public static void getSimilarProductsGcs(
+    String productSetId,
+    String productCategory,
+    String gcsUri,
+    String filter)
+    throws Exception {
+    try (ImageAnnotatorClient queryImageClient = ImageAnnotatorClient.create()) {
+
+        // Get the full path of the product set.
+        String productSetPath = ProductSetName.of(projectId, computeRegion, productSetId).toString();
+
+        // Get the image from Google Cloud Storage
+        ImageSource source = ImageSource.newBuilder().setGcsImageUri(gcsUri).build();
+
+        // Create annotate image request along with product search feature.
+        Feature featuresElement = Feature.newBuilder().setType(Feature.Type.PRODUCT_SEARCH).build();
+        Image image = Image.newBuilder().setSource(source).build();
+        ImageContext imageContext =
+            ImageContext.newBuilder()
+                .setProductSearchParams(
+                    ProductSearchParams.newBuilder()
+                        .setProductSet(productSetPath)
+                        .addProductCategories(productCategory)
+                        .setFilter(filter))
+                .build();
+
+        AnnotateImageRequest annotateImageRequest =
+            AnnotateImageRequest.newBuilder()
+                .addFeatures(featuresElement)
+                .setImage(image)
+                .setImageContext(imageContext)
+                .build();
+        List<AnnotateImageRequest> requests = Arrays.asList(annotateImageRequest);
+
+        // Search products similar to the image.
+        BatchAnnotateImagesResponse response = queryImageClient.batchAnnotateImages(requests);
+
+        List<ProductSearchResults.Result> similarProducts =
+            response.getResponses(0).getProductSearchResults().getResultsList();
```

Comment:
Can you add a comment about why you're only using the first entry in the responses?

Reply:
Nit: it's typical for a comment to come above the line that it's documenting.